File: app/src/main/java/com/qust/lesson/view/LessonRenderData.java
package com.qust.lesson.view;

import android.graphics.Paint;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.qust.lesson.Lesson;
import com.qust.lesson.LessonGroup;
import com.qust.lesson.LessonTable;

import java.util.Arrays;
import java.util.Calendar;

/**
 * 课程渲染信息
 */
public class LessonRenderData{
	
	/**
	 * 开学日期
	 */
	public Calendar startDate;
	
	public LessonHolder[][] lessons;
	
	private int cellWidth;
	
	private Paint textPaint;
	
	private LessonTable lessonTable;
	
	/**
	 * 隐藏教师
	 */
	private boolean hideTeacher;
	
	
	public LessonRenderData(boolean hideTeacher, Paint textPaint){
		this.hideTeacher = hideTeacher;
		this.textPaint = textPaint;
		startDate = Calendar.getInstance();
		lessons = new LessonHolder[7][10];
	}
	
	public void setLessonTable(LessonTable lessonTable){
		this.lessonTable = lessonTable;
	}
	
	public void setMeasureData(int cellWidth){
		this.cellWidth = cellWidth;
	}
	
	/**
	 * 计算绘制课程的信息
	 */
	public void calcLessonData(){
		if(lessonTable == null) return;
		
		startDate.setTime(lessonTable.getStartDay());
		
		LessonGroup[][] lessonGroups = lessonTable.getLessons();
		
		int totalWeek = lessonTable.getTotalWeek();
		
		lessons = new LessonHolder[lessonGroups.length][lessonGroups[0].length];
		
		for(int dayOfWeek = 0; dayOfWeek < lessonGroups.length; dayOfWeek++){
			for(int timeSlot = 0; timeSlot < lessonGroups[0].length; timeSlot++){
				LessonGroup lessonGroup = lessonGroups[dayOfWeek][timeSlot];
				if(lessonGroup == null) continue;
				LessonHolder holder = new LessonHolder(lessonGroup, totalWeek);
				lessons[dayOfWeek][timeSlot] = holder;
			}
		}
	}
	
	/**
	 * 从用于渲染的LessonHolder中获取课表中实际的Lesson
	 * @param dayOfWeek
	 * @param timeSlot
	 * @return
	 */
	public Lesson getLessonByHolder(int week, int dayOfWeek, int timeSlot){
		return lessonTable.getLessons()[dayOfWeek][timeSlot].lessons[lessons[dayOfWeek][timeSlot].index[week]];
	}
	
	/**
	 * 字符串分行
	 * @param src 要分行的字符串
	 * @param des 分行后的字符串
	 * @param maxWidth 一行最大的长度
	 * @param desPos 目标数组开始的位置
	 * @param maxLine 最大支持的行数
	 */
	private int splitString(@NonNull String src, String[] des, int maxWidth, int desPos, int maxLine){
		int srcPos = 0;
		int lines = 0;
		int length = src.length();
		for(int i = srcPos; i < length;){
			i = srcPos + textPaint.breakText(src, i, length, true, maxWidth, null);
			
			if(desPos + lines >= des.length) return lines;
			
			des[desPos + lines] = src.substring(srcPos, i);
			srcPos = i;
			
			if(lines++ == maxLine) return lines;
		}
		return lines;
	}
	
	/**
	 * 课程组
	 */
	public class LessonHolder{
		
		public int[] index;
		
		public int[] count;
		public int[] lessonTime;
		
		public LessonData[] lessonData;
		
		public LessonHolder(@NonNull LessonGroup lessonGroup, int totalWeek){
			Lesson[] group = lessonGroup.lessons;
			
			index = new int[totalWeek];
			count = new int[totalWeek];
			lessonTime = new int[totalWeek];
			lessonData = new LessonData[group.length];
			
			Arrays.fill(index, -1);
			
			for(int i = 0, offset = 1; i < lessonData.length; i++, offset <<= 1){
				Lesson lesson = group[i];
				lessonData[i] = new LessonData(lesson);
				long week = 1L;
				for(int j = 0; j < totalWeek; j++, week <<= 1){
					if((lesson.week & week) > 0){
						if(index[j] == -1) index[j] = i;
						lessonTime[j] |= offset;
						count[j]++;
					}
				}
			}
		}
		
		/**
		 * 是否有下一节课
		 */
		public boolean hasNext(int week){
			return count[week] > 1;
		}
		
		/**
		 * 显示下一节课
		 */
		public void next(int week){
			long offset = 1L << (index[week] + 1);
			for(int i = index[week] + 1; i < count[week]; i++, offset <<= 1){
				if((lessonTime[week] & offset) > 0){
					index[week] = i;
					return;
				}
			}
			index[week] = 0;
		}
		
		@Nullable
		public LessonData current(int week){
			if(count[week] == 0) return null;
			return lessonData[index[week]];
		}
		
		/**
		 * 查找最接近当前周会上的课程
		 *
		 * @param week 当前周
		 * @param findAll 查找全部时间
		 */
		@Nullable
		public LessonData findLesson(int week, boolean findAll){
			if(index[week] != -1) return lessonData[index[week]];
			
			// 向后查找课程
			int i = week + 1;
			for(; i < count.length; i++){
				if(lessonTime[i] > 0){
					int pos = Long.numberOfTrailingZeros(lessonTime[i]);
					index[week] = pos;
					return lessonData[pos];
				}
			}
			
			// 向前查找课程
			if(findAll && week > 0){
				i = week - 1;
				for(; i >= 0; i--){
					if(lessonTime[i] > 0){
						int pos = Long.numberOfTrailingZeros(lessonTime[i]);
						index[week] = pos;
						return lessonData[pos];
					}
				}
			}
			return null;
		}
	}
	
	/**
	 * 课程信息
	 */
	public class LessonData{
		
		/**
		 * 课程类型
		 * 0: auto 自动添加的课程
		 * 1: user 用户创建的课程
		 */
		public int type;
		
		/**
		 * 课程长度
		 */
		public int len;
		
		/**
		 * 课程颜色
		 */
		public int color;
		
		/**
		 * 上课周数
		 */
		public long week;
		
		/**
		 * 文本信息的行数
		 */
		public int lines;
		
		/**
		 * 课程文本信息
		 */
		public String[] data;
		
		public LessonData(@NonNull Lesson lesson){
			this.len = lesson.len;
			this.type = lesson.type;
			this.color = lesson.color;
			
			this.week = lesson.week;
			
			data = new String[len * 3 + (hideTeacher ? 1 : 2)];
			int l = data.length;
			
			lines = splitString(lesson.name, data, cellWidth, 0, l - 3);
			lines += splitString(lesson.place, data, cellWidth, lines + 1, l - lines - 1);
			
			if(!hideTeacher){
				lines += splitString(lesson.teacher, data, cellWidth, lines + 2, l - lines);
			}
		}
	}
}

File: app/src/main/java/com/qust/helper/ui/widget/LessonRenderData.kt
package com.qust.helper.ui.widget

import android.graphics.Paint
import com.qust.helper.data.lesson.Lesson
import com.qust.helper.data.lesson.LessonGroup
import com.qust.helper.data.lesson.LessonTable
import java.util.Arrays

/**
 * 课程渲染信息
 */
class LessonRenderData(
	/**
	 * 隐藏教师
	 */
	private val hideTeacher: Boolean,
	private val textPaint: Paint,
	private val cellWidth: Int = 0,
	val lessonTable: LessonTable? = null
) {

	var lessons: Array<Array<LessonHolder?>> = Array(7) { arrayOfNulls(10) }

	/**
	 * 计算绘制课程的信息
	 */
	fun calcLessonData() {
		lessonTable?.let {
			val lessonGroups = it.lessons
			lessons = Array(lessonGroups.size) { arrayOfNulls(lessonGroups[0].size) }
			for(dayOfWeek in lessonGroups.indices) {
				for(timeSlot in lessonGroups[0].indices) {
					lessonGroups[dayOfWeek][timeSlot]?.let{ lessonGroup ->
						lessons[dayOfWeek][timeSlot] = LessonHolder(lessonGroup, it.totalWeek)
					}
				}
			}
		}
	}

	/**
	 * 从用于渲染的LessonHolder中获取课表中实际的Lesson
	 * @param dayOfWeek
	 * @param timeSlot
	 * @return
	 */
	fun getLessonByHolder(week: Int, dayOfWeek: Int, timeSlot: Int): Lesson {
		return lessonTable!!.lessons[dayOfWeek][timeSlot]!!.lessons[lessons[dayOfWeek][timeSlot]!!.index[week]]
	}

	/**
	 * 字符串分行
	 * @param src 要分行的字符串
	 * @param des 分行后的字符串
	 * @param maxWidth 一行最大的长度
	 * @param desPos 目标数组开始的位置
	 * @param maxLine 最大支持的行数
	 */
	private fun splitString(src: String, des: Array<String?>, maxWidth: Int, desPos: Int, maxLine: Int): Int {
		var srcPos = 0
		var lines = 0
		val length = src.length
		var i = 0
		while(i < length) {
			i = srcPos + textPaint.breakText(src, i, length, true, maxWidth.toFloat(), null)
			if(desPos + lines >= des.size) return lines
			des[desPos + lines] = src.substring(srcPos, i)
			srcPos = i
			if(lines++ == maxLine) return lines
		}
		return lines
	}

	/**
	 * 课程组
	 */
	inner class LessonHolder(
		lessonGroup: LessonGroup,
		totalWeek: Int
	) {

		/**
		 * 某一周的这个时间点同时有几节课
		 */
		var lessonCount = IntArray(totalWeek)

		/**
		 * 当前正在展示的是第几节课，从0开始
		 */
		var index = IntArray(totalWeek).also { Arrays.fill(it, -1) }

		/**
		 * 某一周的这个时间点同时有课的课
		 * 5 = 101，表示当第1，3课会上，第2课不会上
		 */
		private var lessonTime = IntArray(totalWeek)

		private var lessonData: Array<LessonData>

		init {
			var offset = 1
			val array = ArrayList<LessonData>(lessonGroup.lessons.size)
			lessonGroup.lessons.forEachIndexed { i, lesson ->
				array.add(LessonData(lesson))
				var week = 1L
				for(j in 0 until totalWeek) {
					if(lesson.week and week > 0) {
						if(index[j] == -1) index[j] = i
						lessonTime[j] = lessonTime[j] or offset
						lessonCount[j]++
					}
					week = week shl 1
				}
				offset = offset shl 1
			}
			lessonData = array.toTypedArray()
		}

		/**
		 * 是否有多节课
		 */
		fun hasNext(week: Int): Boolean {
			return lessonCount[week] > 1
		}

		/**
		 * 显示下一节课
		 */
		fun next(week: Int) {
			val currentIndex = (index[week] + 1)
			var offset = 1 shl currentIndex
			for(i in currentIndex until  lessonCount[week]) {
				if(lessonTime[week] and offset > 0) {
					index[week] = i
					return
				}
				offset = offset shl 1
			}
			index[week] = 0
		}

		fun current(week: Int): LessonData? {
			return if(lessonCount[week] == 0) null else lessonData[index[week]]
		}

		/**
		 * 查找最接近当前周会上的课程
		 *
		 * @param week 当前周
		 * @param findAll 查找全部时间
		 */
		fun findLesson(week: Int, findAll: Boolean): LessonData? {
			if(index[week] != -1) return lessonData[index[week]]

			// 向后查找课程
			for(i in (week + 1) until lessonCount.size) {
				if(lessonTime[i] > 0) {
					val pos = java.lang.Long.numberOfTrailingZeros(lessonTime[i].toLong())
					index[week] = pos
					return lessonData[pos]
				}
			}

			// 向前查找课程
			if(findAll && week > 0) {
				for(i in (week - 1) downTo 0) {
					if(lessonTime[i] > 0) {
						val pos = java.lang.Long.numberOfTrailingZeros(lessonTime[i].toLong())
						index[week] = pos
						return lessonData[pos]
					}
				}
			}
			return null
		}
	}

	/**
	 * 课程信息
	 * @param type 课程类型
	 * @param len 课程长度
	 * @param color 课程颜色
	 */
	inner class LessonData(
		lesson: Lesson,
		val type: Int = lesson.type,
		val len: Int = lesson.len,
		val color: Int = lesson.color,
	) {
		/**
		 * 文本信息的行数
		 */
		var lines: Int

		/**
		 * 课程文本信息
		 */
		var data: Array<String?> = arrayOfNulls(len * 3 + if(hideTeacher) 1 else 2)

		init {
			val l = data.size
			lines = splitString(lesson.name, data, cellWidth, 0, l - 3)
			lines += splitString(lesson.place, data, cellWidth, lines + 1, l - lines - 1)
			if(!hideTeacher) {
				lines += splitString(lesson.teacher, data, cellWidth, lines + 2, l - lines)
			}
		}
	}
}